fetch no longer works with just --branch

It looks like adding `--ref` in https://github.com/gruntwork-io/fetch/pull/79 may have broken the `--branch` parameter.

With `fetch` version `v0.3.9`:

```
gruntwork-install \
  --module-name "gcp-helpers" \
  --repo "https://github.com/gruntwork-io/terraform-google-ci" \
  --branch "gcp-helpers"
```

Works as follows:

```
2021-02-01 14:10:38 [INFO] [gruntwork-install] Installing from gcp-helpers...
2021-02-01 14:10:38 [INFO] [gruntwork-install] Downloading module gcp-helpers from https://github.com/gruntwork-io/terraform-google-ci
Downloading latest commit from branch "gcp-helpers" of https://github.com/gruntwork-io/terraform-google-ci ...
Extracting files from <repo>/modules/gcp-helpers to /tmp/gruntwork-script-modules/gcp-helpers ...
Download and file extraction complete.
2021-02-01 14:10:39 [INFO] [gruntwork-install] Setting GRUNTWORK_INSTALL_TAG to 
2021-02-01 14:10:39 [INFO] [gruntwork-install] Setting GRUNTWORK_INSTALL_BRANCH to gcp-helpers
2021-02-01 14:10:39 [INFO] [gruntwork-install] Executing /tmp/gruntwork-script-modules/gcp-helpers/install.sh 
```

But with `fetch` version `v0.3.13`:

```
2021-02-01 13:31:02 [INFO] [gruntwork-install] Installing from gcp-helpers...
2021-02-01 13:31:02 [INFO] [gruntwork-install] Downloading module gcp-helpers from https://github.com/gruntwork-io/terraform-google-ci
Downloading git reference "v0.1.0" of https://github.com/gruntwork-io/terraform-google-ci ...
Extracting files from <repo>/modules/gcp-helpers to /tmp/gruntwork-script-modules/gcp-helpers ... 0 files extracted
Download and file extraction complete.
2021-02-01 13:31:02 [ERROR] [gruntwork-install] No files were downloaded. Are you sure "gcp-helpers" is a valid Script Module in https://github.com/gruntwork-io/terraform-google-ci (tag = , branch = gcp-helpers)?
```

I'm guessing [the way `GitRef` is set here to `latestTag`](https://github.com/gruntwork-io/fetch/pull/79/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R285-R295) is causing the code to go down the wrong code path.

--- a/file.go
+++ b/file.go
@@ -151,6 +151,9 @@ func MakeGitHubZipFileRequest(gitHubCommit GitHubCommit, gitHubToken string, ins
 
 	// This represents either a commit, branch, or git tag
 	var gitRef string
+	// Ordering matters in this conditional
+	// GitRef needs to be the fallback and therefore must be last
+	// See https://github.com/gruntwork-io/fetch/issues/87 for an example
 	if gitHubCommit.CommitSha != "" {
 		gitRef = gitHubCommit.CommitSha
 	} else if gitHubCommit.BranchName != "" {

--- a/file.go
+++ b/file.go
@@ -151,14 +151,14 @@ func MakeGitHubZipFileRequest(gitHubCommit GitHubCommit, gitHubToken string, ins
 
 	// This represents either a commit, branch, or git tag
 	var gitRef string
-	if gitHubCommit.GitRef != "" {
-		gitRef = gitHubCommit.GitRef
-	} else if gitHubCommit.CommitSha != "" {
+	if gitHubCommit.CommitSha != "" {
 		gitRef = gitHubCommit.CommitSha
 	} else if gitHubCommit.BranchName != "" {
 		gitRef = gitHubCommit.BranchName
 	} else if gitHubCommit.GitTag != "" {
 		gitRef = gitHubCommit.GitTag
+	} else if gitHubCommit.GitRef != "" {
+		gitRef = gitHubCommit.GitRef
 	} else {
 		return request, fmt.Errorf("Neither a GitCommitSha nor a GitTag nor a BranchName were specified so impossible to identify a specific commit to download.")
 	}

--- a/integration_test.go
+++ b/integration_test.go
@@ -1,0 +1,138 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	cli "gopkg.in/urfave/cli.v1"
+)
+
+func TestFetchWithBranchOption(t *testing.T) {
+	t.Parallel()
+
+	tmpDownloadPath := createTempDir(t, "fetch-branch-test")
+
+	cases := []struct {
+		name       string
+		repoUrl    string
+		branchName string
+		sourcePath string
+	}{
+		// Test on a public repo whose sole purpose is to be a test fixture for this tool
+		{"branch option with public repo", "https://github.com/gruntwork-io/fetch-test-public", "sample-branch", "/"},
+
+		// Private repo equivalent
+		{"branch option with private repo", "https://github.com/gruntwork-io/fetch-test-private", "sample-branch", "/"},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			cmd := fmt.Sprintf("fetch --repo %s --branch %s --source-path %s %s", tc.repoUrl, tc.branchName, tc.sourcePath, tmpDownloadPath)
+			//t.Fatal(cmd)
+			output, _, err := runFetchCommandWithOutput(t, cmd)
+			require.NoError(t, err)
+
+			// When --branch is specified, ensure the latest commit is fetched
+			assert.Contains(t, output, "Downloading latest commit from branch")
+		})
+	}
+}
+
+func runFetchCommandWithOutput(t *testing.T, command string) (string, string, error) {
+	// Note: As most of fetch writes directly to stdout and stderr using the fmt package, we need to temporarily override
+	// the OS pipes. This is based loosely on https://stackoverflow.com/questions/10473800/in-go-how-do-i-capture-stdout-of-a-function-into-a-string/10476304#10476304
+	stdout := bytes.Buffer{}
+	stderr := bytes.Buffer{}
+
+	stdoutReader, stdoutWriter, err1 := os.Pipe()
+	if err1 != nil {
+		return "", "", err1
+	}
+
+	stderrReader, stderrWriter, err2 := os.Pipe()
+	if err2 != nil {
+		return "", "", err2
+	}
+
+	oldStdout := os.Stdout
+	oldStderr := os.Stderr
+
+	// override the pipes to capture output
+	os.Stdout = stdoutWriter
+	os.Stderr = stderrWriter
+
+	// execute the fetch command which produces output
+	err := runFetchCommand(t, command)
+	if err != nil {
+		return "", "", err
+	}
+
+	// copy the output to the buffers in seperate goroutines so printing can't block indefinitely
+	stdoutC := make(chan bytes.Buffer)
+	stderrC := make(chan bytes.Buffer)
+	go func() {
+		var buf bytes.Buffer
+		io.Copy(&buf, stdoutReader)
+		stdoutC <- buf
+	}()
+
+	go func() {
+		var buf bytes.Buffer
+		io.Copy(&buf, stderrReader)
+		stderrC <- buf
+	}()
+
+	// reset the pipes back to normal
+	stdoutWriter.Close()
+	stderrWriter.Close()
+	os.Stdout = oldStdout
+	os.Stderr = oldStderr
+	stdout = <-stdoutC
+	stderr = <-stderrC
+
+	// log the buffers for easier debugging. this is inspired by the integration tests in Terragrunt.
+	// For more information, see: https://github.com/gruntwork-io/terragrunt/blob/master/test/integration_test.go.
+	logBufferContentsLineByLine(t, stdout, "stdout")
+	logBufferContentsLineByLine(t, stderr, "stderr")
+	return stdout.String(), stderr.String(), nil
+}
+
+func runFetchCommand(t *testing.T, command string) error {
+	args := strings.Split(command, " ")
+
+	app := CreateFetchCli(VERSION)
+	app.Action = runFetchWrapper
+	return app.Run(args)
+}
+
+func logBufferContentsLineByLine(t *testing.T, out bytes.Buffer, label string) {
+	t.Logf("[%s] Full contents of %s:", t.Name(), label)
+	lines := strings.Split(out.String(), "\n")
+	for _, line := range lines {
+		t.Logf("[%s] %s", t.Name(), line)
+	}
+}
+
+func createTempDir(t *testing.T, prefix string) string {
+	dir, err := ioutil.TempDir(os.TempDir(), prefix)
+	if err != nil {
+		t.Fatalf("Could not create temporary directory due to error: %v", err)
+	}
+	defer os.RemoveAll(dir)
+	return dir
+}
+
+// We want to call runFetch() using the app.Action wrapper like the main CLI handler, but we don't want to write to strerr
+// and suddenly exit using os.Exit(1), so we use a separate wrapper method in the integration tests.
+func runFetchTestWrapper(c *cli.Context) error {
+	return runFetch(c)
+}

--- a/integration_test.go
+++ b/integration_test.go
@@ -15,34 +15,40 @@ import (
 )
 
 func TestFetchWithBranchOption(t *testing.T) {
-	t.Parallel()
+	// Note: we don't run these tests in parallel as runFetchCommandWithOutput currently overrides the
+	// stdout and stderr variables which may result in unstable tests.
 
 	tmpDownloadPath := createTempDir(t, "fetch-branch-test")
 
 	cases := []struct {
-		name       string
-		repoUrl    string
-		branchName string
-		sourcePath string
+		name         string
+		repoUrl      string
+		branchName   string
+		sourcePath   string
+		expectedFile string
 	}{
 		// Test on a public repo whose sole purpose is to be a test fixture for this tool
-		{"branch option with public repo", "https://github.com/gruntwork-io/fetch-test-public", "sample-branch", "/"},
+		{"branch option with public repo", "https://github.com/gruntwork-io/fetch-test-public", "sample-branch", "/", "foo.txt"},
 
 		// Private repo equivalent
-		{"branch option with private repo", "https://github.com/gruntwork-io/fetch-test-private", "sample-branch", "/"},
+		{"branch option with private repo", "https://github.com/gruntwork-io/fetch-test-private", "sample-branch", "/", "bar.txt"},
 	}
 
 	for _, tc := range cases {
+		// The following is necessary to make sure tc's values don't
+		// get updated due to concurrency within the scope of t.Run(..) below
 		tc := tc
 
 		t.Run(tc.name, func(t *testing.T) {
 			cmd := fmt.Sprintf("fetch --repo %s --branch %s --source-path %s %s", tc.repoUrl, tc.branchName, tc.sourcePath, tmpDownloadPath)
-			//t.Fatal(cmd)
 			output, _, err := runFetchCommandWithOutput(t, cmd)
 			require.NoError(t, err)
 
 			// When --branch is specified, ensure the latest commit is fetched
 			assert.Contains(t, output, "Downloading latest commit from branch")
+
+			// Ensure the expected file was downloaded
+			assert.FileExists(t, JoinPath(tmpDownloadPath, tc.expectedFile))
 		})
 	}
 }
@@ -50,6 +56,7 @@ func TestFetchWithBranchOption(t *testing.T) {
 func runFetchCommandWithOutput(t *testing.T, command string) (string, string, error) {
 	// Note: As most of fetch writes directly to stdout and stderr using the fmt package, we need to temporarily override
 	// the OS pipes. This is based loosely on https://stackoverflow.com/questions/10473800/in-go-how-do-i-capture-stdout-of-a-function-into-a-string/10476304#10476304
+	// Our goal eventually is to remove this, but we'll need to introduce a logger as mentioned in: https://github.com/gruntwork-io/fetch/issues/89
 	stdout := bytes.Buffer{}
 	stderr := bytes.Buffer{}
 
@@ -76,7 +83,7 @@ func runFetchCommandWithOutput(t *testing.T, command string) (string, string, er
 		return "", "", err
 	}
 
-	// copy the output to the buffers in seperate goroutines so printing can't block indefinitely
+	// copy the output to the buffers in separate goroutines so printing can't block indefinitely
 	stdoutC := make(chan bytes.Buffer)
 	stderrC := make(chan bytes.Buffer)
 	go func() {

--- a/main.go
+++ b/main.go
@@ -52,12 +52,19 @@ const optionWithProgress = "progress"
 
 const envVarGithubToken = "GITHUB_OAUTH_TOKEN"
 
-func main() {
+// Create the Fetch CLI App
+func CreateFetchCli(version string) *cli.App {
+	cli.OsExiter = func(exitCode int) {
+		// Do nothing. We just need to override this function, as the default value calls os.Exit, which
+		// kills the app (or any automated test) dead in its tracks.
+	}
+
 	app := cli.NewApp()
 	app.Name = "fetch"
 	app.Usage = "fetch makes it easy to download files, folders, and release assets from a specific git commit, branch, or tag of public and private GitHub repos."
 	app.UsageText = "fetch [global options] <local-download-path>\n   (See https://github.com/gruntwork-io/fetch for examples, argument definitions, and additional docs.)"
-	app.Version = VERSION
+	app.Author = "Gruntwork <www.gruntwork.io>"
+	app.Version = version
 
 	app.Flags = []cli.Flag{
 		cli.StringFlag{
@@ -112,6 +119,11 @@ func main() {
 		},
 	}
 
+	return app
+}
+
+func main() {
+	app := CreateFetchCli(VERSION)
 	app.Action = runFetchWrapper
 
 	// Run the definition of App.Action

--- a/main.go
+++ b/main.go
@@ -302,6 +302,9 @@ func downloadSourcePaths(sourcePaths []string, destPath string, githubRepo GitHu
 
 	// Download that release as a .zip file
 
+	// Ordering matters in this conditional
+	// GitRef needs to be the fallback and therefore must be last
+	// See https://github.com/gruntwork-io/fetch/issues/87 for an example
 	if gitHubCommit.CommitSha != "" {
 		fmt.Printf("Downloading git commit \"%s\" of %s ...\n", gitHubCommit.CommitSha, githubRepo.Url)
 	} else if gitHubCommit.BranchName != "" {

--- a/main.go
+++ b/main.go
@@ -290,14 +290,14 @@ func downloadSourcePaths(sourcePaths []string, destPath string, githubRepo GitHu
 
 	// Download that release as a .zip file
 
-	if gitHubCommit.GitRef != "" {
-		fmt.Printf("Downloading git reference \"%s\" of %s ...\n", gitHubCommit.GitRef, githubRepo.Url)
-	} else if gitHubCommit.CommitSha != "" {
+	if gitHubCommit.CommitSha != "" {
 		fmt.Printf("Downloading git commit \"%s\" of %s ...\n", gitHubCommit.CommitSha, githubRepo.Url)
 	} else if gitHubCommit.BranchName != "" {
 		fmt.Printf("Downloading latest commit from branch \"%s\" of %s ...\n", gitHubCommit.BranchName, githubRepo.Url)
 	} else if gitHubCommit.GitTag != "" {
 		fmt.Printf("Downloading tag \"%s\" of %s ...\n", latestTag, githubRepo.Url)
+	} else if gitHubCommit.GitRef != "" {
+		fmt.Printf("Downloading git reference \"%s\" of %s ...\n", gitHubCommit.GitRef, githubRepo.Url)
 	} else {
 		return fmt.Errorf("The commit sha, tag, and branch name are all empty.")
 	}

--- a/util.go
+++ b/util.go
@@ -1,0 +1,10 @@
+package main
+
+import "path/filepath"
+
+// Windows systems use \ as the path separator *nix uses /
+// Use this function when joining paths to force the returned path to use / as the path separator
+// This will improve cross-platform compatibility
+func JoinPath(elem ...string) string {
+	return filepath.ToSlash(filepath.Join(elem...))
+}

